lib/index.js

```diff
@@ -1,53 +1,80 @@
+const bodyParser = require('koa-bodyparser');
+const conditional = require('koa-conditional-get');
+const convert = require('koa-convert');
+const CSRF = require('koa-csrf');
+const error = require('koa-error');
+const etag = require('koa-etag');
+const Koa = require('koa');
+const merge = require('merge-descriptors');
+const Pug = require('koa-pug')
+const Qs = require('koa-qs');
+const trace = require('koa-trace');
 
-var koa = require('koa');
-var merge = require('merge-descriptors');
-
-var middleware = require('./middleware');
-var ctx = require('./context');
-var req = require('./request');
-var res = require('./response');
-var App = require('./app');
-
-module.exports = function koala(options) {
-  options = options || {};
-
-  var app = koa();
-
-  middleware(app, options);
-  merge(app.context, ctx);
-  merge(app.request, req);
-  merge(app.response, res);
-  App(app);
-
-  // handlers
-  app.errorHandler = require('koa-error')(options.error);
-  app.pageNotFoundHandler = pageNotFoundHandler;
-
-  // properties
-  app.jsonStrict = options.jsonStrict !== false;
-
-  // proto stuff
-  require('koa-csrf')(app);
-  require('koa-trace')(app);
-  require('koa-body-parsers')(app);
-  // nested query string support
-  if (options.qs) {
-    require('koa-qs')(app);
-    app.querystring = require('qs');
-  }
+const middleware = require('./middleware');
+const ctx = require('./context');
+const req = require('./request');
+const res = require('./response');
+const App = require('./app');
 
-  // jsonp support
-  if (options.jsonp) {
-    require('koa-safe-jsonp')(app, options.jsonp);
-  }
+module.exports = class Koala extends Koa {
+  constructor(options = {}) {
+    super();
+
+    middleware(this, options);
+    merge(this.context, ctx);
+    merge(this.request, req);
+    merge(this.response, res);
+    App(this);
+
+    // handlers
+    this.pageNotFoundHandler = pageNotFoundHandler;
 
-  if (process.env.NODE_ENV !== 'production') app.debug();
+    // properties
+    this.jsonStrict = options.jsonStrict !== false;
 
-  return app;
+    // proto stuff
+    this.use(conditional());
+    this.use(new CSRF({
+      invalidSessionSecretMessage: 'Invalid session secret',
+      invalidSessionSecretStatusCode: 403,
+      invalidTokenMessage: 'Invalid CSRF token',
+      invalidTokenStatusCode: 403,
+      excludedMethods: [ 'GET', 'HEAD', 'OPTIONS' ],
+      disableQuery: false
+    }));
+    this.use(etag());
+    this.use(bodyParser());
+    // this.use(convert(trace));
+    const pug = new Pug({
```

Comment:
I will prefer passing a merged `options.pug` object  to pug, something like `new Pug(options.pug)` with required defaults.
So that it can be configured by someone who is using this.